[java-generator] - Gradle plugin

fabiobrz opened the pull request:
## Description
Adding a basic implementation for a Java Generator Gradle plugin

FYI @andreaTP 

Fix #4829 

## Type of change
<!---
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [x] I tested my code in Kubernetes
 - [x] I tested my code in OpenShift

<!--
Integration tests (https://github.com/fabric8io/kubernetes-client/tree/master/kubernetes-itests)
Please check integration tests and provide/improve tests if applicable.

Open your PR in Draft mode and verify all of the applicable Checklist items before marking your pull request as ready for review
-->

andreaTP commented on java-generator/it/src/it/plugin/gradle/simple/build.gradle:
```diff
@@ -0,0 +1,31 @@
+
+javaGen {
+    source = file('artemis-crd.yml')
+    target = file('tmp/generated/')
```
Question, is the `target` folder already included in the "Source folders" of Gradle?

fabiobrz replied:
I am not sure to get the gist here: is your questio about the `target` property key or rather about its value? The reason I am asking: do we want to changethe default _value_ or the test one to something like `build/generated/` or do we want to chenge the _key_ to something in the Gradle semantics? 
If the latter is the case we should ask ourselves whether we agree on having configuration key names to differe from the ones of the Maven plugin.
WDYT?

andreaTP replied:
Sorry for the confusion, I'm referring to the "default value".

fabiobrz replied:
This should be fixed now

andreaTP commented on java-generator/it/src/test/java/io/fabric8/java/generator/gradle/plugin/ITGradleRunnerExtension.java:
```diff
@@ -0,0 +1,72 @@
+  @Override
+  public void beforeEach(ExtensionContext context) throws Exception {
+    gradleRunner = GradleRunner.create()
+        .withGradleDistribution(new URI("https://services.gradle.org/distributions/gradle-8.1-bin.zip"))
```
would it be possible to bind the gradle version to a Maven property so that it will be updated by dependabot?

fabiobrz replied:
Yep. I think this is doable by adding a `*.properties` file that binds a property value to one defined by the Maven project that's executing the tests.
BTW - I must say this would diverge from the version of the Gradle artifacts we're using, which is behind ATM (see https://github.com/fabiobrz/kubernetes-client/blob/java-gen.gradle-plugin/java-generator/pom.xml#L43) and which doesn't seemm to be actively released lately, see e.g. `gradle-core` [1].
WDYT?

[1]
https://mvnrepository.com/artifact/org.gradle/gradle-core?repo=gradle-libs-releases-local

andreaTP replied:
You should be able to skip the `properties` file and directly generate from Maaven a Java source file containing the version, we usually use something like this: https://stackoverflow.com/a/36628755

Regarding the gradle versions issues, I'm not really familiar with them ... maybe @manusa ?

fabiobrz replied:
Should be fixed now. Currently the Gradle binaries version is set by default to `8.1`. Can be overridden via `-Dgradle-binaries.version=...`

andreaTP replied:
perfect 👍

andreaTP commented on java-generator/gradle-plugin/src/main/java/io/fabric8/java/generator/gradle/plugin/JavaGeneratorPluginExtension.java:
```diff
@@ -0,0 +1,128 @@
+
+  public File getDownloadTargetOrDefault() {
+    return this.getDownloadTarget().getAsFile()
+        .getOrElse(this.gradleProject.getLayout().getProjectDirectory().dir("tmp").dir("input").getAsFile());
```
is it "gradle idiomatic" to create a `tmp` folder at the root of the project?
I'm not sure I have seen this pattern before ...

fabiobrz replied:
No, you're right, thanks for pointing this out. We could set this default differenlty. I'll have a look.

andreaTP replied:
thanks 👍

fabiobrz replied:
This should be fixed now.

andreaTP commented on java-generator/it/src/it/plugin/gradle/simple/expected/ActiveMQArtemisSpec.expected:
```diff
@@ -0,0 +1,169 @@
+package io.amq.broker.v1beta1;
```
instead of checking the content of the generated files would it be possible to verify in a test that the generated sources have been compiled and are available on the classpath?

fabiobrz replied:
This should be fixed now.

andreaTP commented on java-generator/gradle-plugin/src/main/java/io/fabric8/java/generator/gradle/plugin/JavaGeneratorPluginExtension.java:
```diff
@@ -0,0 +1,128 @@
+   * Generate uppercase Enums
+   *
+   */
+  public abstract Property<Boolean> getEnumUppercase();
```
Instead of flattening the `Config` down here, I think it's more convenient (especially for maintenance) to re-use a private instance of Config: https://github.com/fabric8io/kubernetes-client/blob/master/java-generator/core/src/main/java/io/fabric8/java/generator/Config.java
so that we can maintain defaults etc. etc. in a single place.

fabiobrz replied:
Fixed, thanks.

andreaTP reviewed (commented):
Outstanding work @fabiobrz , thanks a ton!
Even an integration test! ❤️ 

I added a few questions/notes, especially due to my inexperience with Gradle plugins.
I think that @manusa input would be helpful here 🙂

andreaTP commented on java-generator/gradle-plugin/pom.xml:
```diff
@@ -0,0 +1,122 @@
+    </dependency>
+    <dependency>
+      <groupId>org.gradle</groupId>
+      <artifactId>gradle-base-services</artifactId>
```
Forgot to mention, but I think that most of these dependencies should be marked as `provided`.

fabiobrz replied:
Yep, I noticed that is the case in Jkube. I'll have a look again and fix it, thanks!

andreaTP commented on java-generator/gradle-plugin/src/main/java/io/fabric8/java/generator/gradle/plugin/JavaGeneratorPluginExtension.java:
```diff
@@ -33,10 +31,20 @@ public abstract class JavaGeneratorPluginExtension {
   public static final String NAME = "javaGen";
   private Project gradleProject;
 
+  private Config javaGeneratorConfig = new Config(
```
```suggestion
  private Config javaGeneratorConfig = Config.builder().build();
```

fabiobrz replied:
Yep, done!

andreaTP commented on java-generator/it/src/test/java/io/fabric8/java/generator/gradle/plugin/SimpleIT.java:
```diff
@@ -0,0 +1,119 @@
+  @RegisterExtension
+  private final ITGradleRunnerExtension gradleRunner = new ITGradleRunnerExtension();
+
+  static final List<File> TESTED_CRDS = Stream.of(
```
In other tests we are using the `junit` `TempDir`, e.g.:
https://github.com/fabric8io/kubernetes-client/blob/fe579f0eab6ece572cee8a184f0614864a8c95a6/java-generator/core/src/test/java/io/fabric8/java/generator/CompilationTest.java#L45-L46

would it be possible to apply the same pattern?

andreaTP replied:
applies at line 45, sorry for the confusion.

fabiobrz replied:
I am wondering whether this is the right use case for `@TempDir`. Maybe I am missing some pieces but according to the docs its lifecycle would be managed by JUnit. Which clashes with the fact that those directories are instead managed by the Gradle build. `resources` is an exception, but I'll use it to store the CRD locally.

andreaTP replied:
Oh, if those are managed by Gradle it's fine, missed it 👍

andreaTP commented on java-generator/gradle-plugin/src/main/java/io/fabric8/java/generator/gradle/plugin/JavaGeneratorPluginExtension.java:
```diff
@@ -0,0 +1,156 @@
+  public File getDownloadTargetOrDefault() {
+    return this.getDownloadTarget().getAsFile()
+        .getOrElse(this.gradleProject.getLayout().getProjectDirectory()
+            .dir("src")
```
Usually, we download the CRD to a "temp" or "target" dir instead of a "managed source folder", something like:
`build/crds` wdyt?

fabiobrz replied:
`build` would be in the Gradle lifecycle managment scope. If we want to use a default here - then either we live with a Java project structure and let the current structure be or we use something different, e.g.: `tmp/crds`, although this would mean for us to ger back to my initial implementation, basically.
TBH, we're generating Java sources, which to my eyes makes the current implementation legit, rather than the proposed one. I can be proved wrong, thoug, just let me know what you think.

fabiobrz replied:
Now I got what you mean, i.e. putput the downloaded files to the build folder, so calling `clean` would... clean :slightly_smiling_face: 
Done, thanks!

andreaTP commented on java-generator/it/src/it/plugin/gradle/simple/build.gradle:
```diff
@@ -0,0 +1,49 @@
+
+javaGen {
+    urls = [
+      "https://raw.githubusercontent.com/artemiscloud/activemq-artemis-operator/v1.0.11/deploy/crds/broker_activemqartemis_crd.yaml",
```
So far, we have attempted to keep the build and test "local", would it be possible to commit one CRD and test against it?

fabiobrz replied:
Yep, done!

andreaTP commented on java-generator/it/src/main/java-templates/io/fabric8/java/generator/testing/GradleProperties.java:
```diff
@@ -0,0 +1,25 @@
+ */
+package io.fabric8.java.generator.testing;
+
+public final class GradleProperties {
```
👍

andreaTP reviewed (changes requested):
A last round of minor comments and is good to go! Thanks again @fabiobrz !

fabiobrz commented:
Hi @andreaTP - I added one unit test, at least for the plugin class itself, and some documentation. I hope I addressed all the concerns you pointed out. Marking the PR as ready.

Feel free to let me know in case you have additional thoughts.

andreaTP reviewed (approved):
Amazing work LGTM, thanks a lot @fabiobrz !

manusa reviewed (approved):
LGTM, thx!

manusa reviewed (approved):
LGTM, thx!
